llvm/lib/Transforms/Vectorize/VectorCombine.cpp

```diff
@@ -929,6 +930,158 @@ bool VectorCombine::foldBitOpOfCastops(Instruction &I) {
   return true;
 }
 
+struct PreservedCastFlags {
+  bool NNeg = false;
+  bool NUW = false;
+  bool NSW = false;
+};
+
+// Try to cast C to InvC losslessly, satisfying CastOp(InvC) == C.
+// Will try best to preserve the flags.
+static Constant *getLosslessInvCast(Constant *C, Type *InvCastTo,
```

Comment:
Maybe we could put this in `InstCombineInternal.h` where we have more methods like `getLosslessUnsignedTrunc`, `getLosslessSignedTrunc`, etc. ?

Reply:
As this is based on ConstantFolding APIs, maybe ConstantFolding.h may be a good place?